README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

<!-- badges: start -->

[![](https://cranlogs.r-pkg.org/badges/grand-total/pct)](https://cran.r-project.org/package=pct)
[![R build
status](https://github.com/itsleeds/pct/workflows/R-CMD-check/badge.svg)](https://github.com/itsleeds/pct/actions)
[![](https://www.r-pkg.org/badges/version/pct)](https://www.r-pkg.org/pkg/pct)
[![R-CMD-check](https://github.com/itsleeds/pct/workflows/R-CMD-check/badge.svg)](https://github.com/itsleeds/pct/actions)
<!-- badges: end -->

<!-- [![R build status](https://github.com/itsleeds/pct/workflows/R-CMD-check/badge.svg)](https://github.com/itsleeds/pct/actions) -->

<!-- [![Travis build status](https://travis-ci.org/ITSLeeds/pct.svg?branch=master)](https://travis-ci.org/ITSLeeds/pct) -->

# pct

The goal of pct is to make the data produced by the Propensity to Cycle
Tool (PCT) easier to access and reproduce. The PCT a research project
and web application hosted at [www.pct.bike](https://www.pct.bike/). For
an overview of the data provided by the PCT, clicking on the previous
link and trying it out is a great place to start. An academic
[paper](https://www.jtlu.org/index.php/jtlu/article/view/862) on the PCT
provides detail on the motivations for and methods underlying the
project.

A major motivation behind the project was making transport evidence more
accessible, encouraging evidence-based transport policies. The code base
underlying the PCT is publicly available (see
[github.com/npct](https://github.com/npct/)). However, the code hosted
there is not easy to run or reproduce, which is where this package comes
in: it provides quick access to the data underlying the PCT and enables
some of the key results to be reproduced quickly. It was developed
primarily for educational purposes (including for upcoming PCT training
courses) but it may be useful for people to build on the the methods,
for example to create a scenario of cycling uptake in their
town/city/region.

In summary, if you want to know how PCT works, be able to reproduce some
of its results, and build scenarios of cycling uptake to inform
transport policies enabling cycling in cities worldwide, this package is
for you\!

## Installation

``` r
# from CRAN
install.packages("pct")
```

You can install the development version of the package as follows:

``` r
remotes::install_github("ITSLeeds/pct")
```

Load the package as follows:

``` r
library(pct)
```

## Documentation

Probably the best place to get further information on the PCT is from
the package’s website at <https://itsleeds.github.io/pct/>

There you will find the following vignettes, which we recommend reading,
and reproducing and experimenting with the code contained within to
deepen your understanding of the code, in the following order:

1.  A ‘get started’ introduction to the PCT and associated R package:
    <https://itsleeds.github.io/pct/articles/pct.html>
2.  Getting and using PCT data, an
    [article](https://itsleeds.github.io/pct/articles/getting.html)
    showing how to get and use data from the PCT, based on a case study
    from North Yorkshire
3.  A [training
    vignette](https://itsleeds.github.io/pct/articles/pct_training.html)
    providing more detailed guidance on data provided by the PCT
    package, with interactive exercises based on a case study of the
    Isle of Wight
4.  A
    [vignette](https://itsleeds.github.io/pct/articles/cycling-potential-uk.html)
    show how to use the data provided by the package to estimate cycling
    uptake in UK cities
5.  A
    [vignette](https://itsleeds.github.io/pct/articles/pct-international.html)
    demonstrating the international applicability of the PCT method,
    with help from this and other R packages

You will also find there documentation for each of the functions at
[itsleeds.github.io/pct/reference/](https://itsleeds.github.io/pct/reference/index.html).
Below we describe some of the basics.

## Get PCT data

From feedback, we hear that the use of the data is critical in decision
making. Therefore, one area where the package could be useful is making
the data “easily” available to be processed.

  - `get_pct`: the basic function to obtain data available
    [here](https://itsleeds.github.io/pct/reference/get_pct.html).

The rest of these should be self explanatory.

  - `get_pct_centroids`
  - `get_pct_lines`
  - `get_pct_rnet`
  - `get_pct_routes_fast`
  - `get_pct_routes_quiet`
  - `get_pct_zones`
  - `uptake_pct_godutch`
  - `uptake_pct_govtarget`

For example, to get the centroids in West Yorkshire:

``` r
centroids = get_pct_centroids(region = "west-yorkshire")
plot(centroids[, "geo_name"])
```

<img src="man/figures/README-centroids-1.png" width="100%" />

Likewise to download the desire lines for “west-yorkshire”:

``` r
lines = get_pct_lines(region = "west-yorkshire")
lines = lines[order(lines$all, decreasing = TRUE), c("all")]
plot(lines[1:10,], lwd = 4)
```

<img src="man/figures/README-get_pct_lines-1.png" width="100%" />

``` r
# view the lines on a map
# mapview::mapview(lines[1:3000, c("geo_name1")])
```

## Estimate cycling uptake

An important part of the PCT is its ability to create model scenarios of
cycling uptake. Key to the PCT uptake model is ‘distance decay’, meaning
that short trips are more likely to be cycled than long trips. The
functions `uptake_pct_govtarget()` and `uptake_pct_godutch()` implement
uptake models used in the PCT, which use distance and hilliness per
desire line as inputs and output the proportion of people who could be
expected to cycle if that scenario were realised. The scenarios of
cycling uptake produced by these functions are not predictions of what
*will* happen, but illustrative snapshots of what *could* happen if
overall propensity to cycle reached a certain level. The uptake levels
produced by Go Dutch and Government Target scenarios (which represent
increases in cycling, not final levels) are illustrated in the graph
below (other scenarios could be produced, see the [source
code](https://itsleeds.github.io/pct/reference/uptake_pct_govtarget.html)
see how these models work):

``` r
max_distance = 50
distances = 1:max_distance
max_hilliness = 5
hilliness = 0:max_hilliness
uptake_df = data.frame(
  distances = rep(distances, times = max_hilliness + 1),
  hilliness = rep(hilliness, each = max_distance)
)
p_govtarget = uptake_pct_govtarget(
    distance = uptake_df$distances,
    gradient = uptake_df$hilliness
    )
p_godutch = uptake_pct_godutch(
    distance = uptake_df$distances,
    gradient = uptake_df$hilliness
    )
uptake_df = rbind(
  cbind(uptake_df, scenario = "govtarget", pcycle = p_govtarget),
  cbind(uptake_df, scenario = "godutch", pcycle = p_godutch)
)
library(ggplot2)
ggplot(uptake_df) +
  geom_line(aes(
    distances,
    pcycle,
    linetype = scenario,
    colour = as.character(hilliness)
  )) +
  scale_color_discrete("Gradient (%)")
```

<img src="man/figures/README-decay-1.png" width="100%" />

The proportion of trips made by cycling along each origin-destination
(OD) pair therefore depends on the trip distance and hilliness. The
equivalent plot for hilliness is as follows:

``` r
distances = c(1, 3, 6, 10, 15, 21)
hilliness = seq(0, 10, by = 0.2)
uptake_df = 
  data.frame(
    expand.grid(distances, hilliness)
  )
names(uptake_df) = c("distances", "hilliness")
p_govtarget = uptake_pct_govtarget(
    distance = uptake_df$distances,
    gradient = uptake_df$hilliness
    )
p_godutch = uptake_pct_godutch(
    distance = uptake_df$distances,
    gradient = uptake_df$hilliness
    )
uptake_df = rbind(
  cbind(uptake_df, scenario = "govtarget", pcycle = p_govtarget),
  cbind(uptake_df, scenario = "godutch", pcycle = p_godutch)
)
ggplot(uptake_df) +
  geom_line(aes(
    hilliness,
    pcycle,
    linetype = scenario,
    colour = formatC(distances, flag = "0", width = 2)
  )) +
  scale_color_discrete("Distance (km)")
```

<img src="man/figures/README-decayhills-1.png" width="100%" />

Note: if distances or gradient values appear to be provided in incorrect
units, they will automatically be updated:

``` r
distances = uptake_df$distances * 1000
hilliness = uptake_df$hilliness / 100
res = uptake_pct_godutch(distances, hilliness, verbose = TRUE)
#> Distance assumed in m, switching to km
#> Gradient assumed to be gradient, switching to % (*100)
```

The main input dataset into the PCT is OD data and, to convert each OD
pair into a geographic desire line, geographic zone or centroids.
Typical input data is provided in packaged datasets `od_leeds` and
`zones_leeds`, as shown in the next section.

## Reproduce PCT for Leeds

This example shows how scenarios of cycling uptake, and how ‘distance
decay’ works (short trips are more likely to be cycled than long trips).

The input data looks like this (origin-destination data and geographic
zone data):

``` r
class(od_leeds)
#> [1] "tbl_df"     "tbl"        "data.frame"
od_leeds[c(1:3, 12)]
#> # A tibble: 10 × 4
#>    area_of_residence area_of_workplace   all bicycle
#>    <chr>             <chr>             <dbl>   <dbl>
#>  1 E02002363         E02006875           922      43
#>  2 E02002373         E02006875          1037      73
#>  3 E02002384         E02006875           966      13
#>  4 E02002385         E02006875           958      52
#>  5 E02002392         E02006875           753      19
#>  6 E02002404         E02006875          1145      10
#>  7 E02002411         E02006875           929      27
#>  8 E02006852         E02006875          1221      99
#>  9 E02006861         E02006875          1177      56
#> 10 E02006876         E02006875          1035      10
class(zones_leeds)
#> [1] "sf"         "data.frame"
zones_leeds[1:3, ]
#> old-style crs object detected; please recreate object with a recent sf::st_crs()
#> old-style crs object detected; please recreate object with a recent sf::st_crs()
#> Simple feature collection with 3 features and 6 fields
#> Geometry type: MULTIPOLYGON
#> Dimension:     XY
#> Bounding box:  xmin: -1.727245 ymin: 53.90046 xmax: -1.294313 ymax: 53.94589
#> Geodetic CRS:  WGS 84
#>      objectid  msoa11cd  msoa11nm msoa11nmw st_areasha st_lengths
#> 2270     2270 E02002330 Leeds 001 Leeds 001    3460674  10002.983
#> 2271     2271 E02002331 Leeds 002 Leeds 002   21870986  26417.665
#> 2272     2272 E02002332 Leeds 003 Leeds 003    2811303   8586.548
#>                            geometry
#> 2270 MULTIPOLYGON (((-1.392046 5...
#> 2271 MULTIPOLYGON (((-1.340405 5...
#> 2272 MULTIPOLYGON (((-1.682211 5...
```

The `stplanr` package can be used to convert the non-geographic OD data
into geographic desire lines as follows:

``` r
library(sf)
#> Linking to GEOS 3.11.1, GDAL 3.6.2, PROJ 9.1.1; sf_use_s2() is TRUE
desire_lines = stplanr::od2line(flow = od_leeds, zones = zones_leeds[2])
#> Creating centroids representing desire line start and end points.
#> old-style crs object detected; please recreate object with a recent sf::st_crs()
#> old-style crs object detected; please recreate object with a recent sf::st_crs()
#> old-style crs object detected; please recreate object with a recent sf::st_crs()
#> old-style crs object detected; please recreate object with a recent sf::st_crs()
#> old-style crs object detected; please recreate object with a recent sf::st_crs()
#> old-style crs object detected; please recreate object with a recent sf::st_crs()
plot(desire_lines[c(1:3, 12)])
```

<img src="man/figures/README-desire-1.png" width="100%" />

We can convert these straight lines into routes with a routing service,
e.g.:

``` r
segments_fast = stplanr::route(l = desire_lines, route_fun = cyclestreets::journey)
#> Most common output is sf
```

We got useful information from this routing operation, we will convert
the route segments into complete routes with `dplyr`:

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
routes_fast = segments_fast %>% 
  group_by(area_of_residence, area_of_workplace) %>% 
  summarise(
    all = unique(all),
    bicycle = unique(bicycle),
    length = sum(distances),
    av_incline = mean(gradient_smooth) * 100
  ) 
#> `summarise()` has grouped output by 'area_of_residence'. You can override using
#> the `.groups` argument.
```

The results at the route level are as follows:

``` r
plot(routes_fast)
```

<img src="man/figures/README-routes_fast-1.png" width="100%" />

Now we estimate cycling uptake:

``` r
routes_fast$uptake = uptake_pct_govtarget(distance = routes_fast$length, gradient = routes_fast$av_incline)
routes_fast$bicycle_govtarget = routes_fast$bicycle +
  round(routes_fast$uptake * routes_fast$all)
```

Let’s see how many people started cycling:

``` r
sum(routes_fast$bicycle_govtarget) - sum(routes_fast$bicycle)
#> [1] 408
```

Nearly 1000 more people cycling to work, just in 10 desire is not bad\!
What % cycling is this, for those routes?

``` r
sum(routes_fast$bicycle_govtarget) / sum(routes_fast$all)
#> [1] 0.07985803
sum(routes_fast$bicycle) / sum(routes_fast$all)
#> [1] 0.03963324
```

It’s gone from 4% to 11%, a realistic increase if cycling were enabled
by good infrastructure and policies.

Now: where to prioritise that infrastructure and those policies?

``` r
routes_fast_linestrings = sf::st_cast(routes_fast, "LINESTRING")
rnet = stplanr::overline(routes_fast_linestrings, attrib = c("bicycle", "bicycle_govtarget"))
lwd = rnet$bicycle_govtarget / mean(rnet$bicycle_govtarget)
plot(rnet["bicycle_govtarget"], lwd = lwd)
```

<img src="man/figures/README-rnetgove-1.png" width="100%" />

We can view the results in an interactive map and share with policy
makers, stakeholders, and the public\! E.g. (see interactive map
[here](https://rpubs.com/RobinLovelace/474074)):

``` r
mapview::mapview(rnet, zcol = "bicycle_govtarget", lwd = lwd * 2)
```

![](https://raw.githubusercontent.com/ITSLeeds/pct/master/pct-leeds-demo.png)

## Limitations

  - This package does not contain code to estimate cycling uptake
    associated with intrazonal flows and people with no fixed job data,
    although the datasets downloaded with the `get_pct_centroids()`
    functions provide estimated uptake for intrazonal flows.
  - This package currently does not contiain code to estimate health
    benefits

<!-- Commented out as we have now immediate plans to work on these -->

<!-- ## Next steps and further resources -->

<!-- - Add additional scenarios of cycling uptake from different places (e.g. goCambridge) -->

<!-- - Add additional distance decay functions -->

<!-- - Make it easy to use data from other cities around the world -->

<!-- - Show how to create raster tiles of cycling uptake -->

## Testing the package

Test the package with the following code:

``` r
remotes::install_github("ITSLeeds/pct")
devtools::check()
```


NEWS.md
# pct 0.9.9 (May 2023)

- Fix for CRAN check errors (pass tests without internet) (#124)
- Update README with test instructions
- Ignore large .geojson files in .Rbuildignore
- Prevent high uptake numbers for long distances (#125)

# pct 0.9.8 (February 2023)

- Remove tidyverse (#123)

# pct 0.9.7

- Additional changes for CRAN (#222)
- Message rather than error if resource unavailable (#222)

# pct 0.9.6

- Updates to fail gracefully (#220)

# pct 0.9.5 (November 2022)

- Updated docs to fix CRAN check errors (#119)

# pct 0.9.4 (November 2021)

- Typos fixed in README

# pct 0.9.3 (November 2021)

- Updated URLs to comply with CRAN policies
- Make package tests quicker

# pct 0.9.2

- New vignette on getting and modelling PCT data

# pct 0.9.1

- Minor updates to vignettes to support `sf` 1.0 (#101)

# pct 0.9.0

- New function `uptake_pct_ebike_2020()`

# pct 0.8.0

- Check and update hilliness if the seem to be in incorrect units, and document (#70)
- New vignette contributed by Nathanael Sheehan

# pct 0.7.0

- New argument `verbose`, false by default, in uptake functions
- Updated training vignette
- Updated introductory `pct` vignette
- Fix in code for checks

# pct 0.6.0

- New functions `uptake_pct_govtarget_school2()` and `uptake_pct_godutch_school2()`
- Enables download of other OD datasets with `get_od()` (#66)

# pct 0.5.0

- New `uptake_pct_govtarget_2020()` and `uptake_pct_godutch_2020()` functions approximate the uptake models used in the updated 2020 release of the PCT results
- New `model_pcycle_pct_2020()` function to estimate cycling levels as a function of distance and hilliness, using the uptake function in the 2020 release of the PCT.
- Fix parameters in `uptake_pct_godutch()` (#65)
- Higher resolution LSOA data now the default (`geography = 'lsoa'`)

# pct 0.4.1

- Update vignette for recent version of `sf` (#64)

# pct 0.4.0

- Read `.geojson` files directly rather than downloading old `.Rds` files that were generating error messages - see https://github.com/ITSLeeds/pct/issues/57
- Do not test time-consuming tests on CRAN as per #58

# pct 0.3.0

- Always returns objects with EPSG code 4326, see https://github.com/ropensci/stats19/issues/135

# pct 0.2.7

- Improved documentation for godutch uptake function
- Uptake function now work when there are NAs in the distances (previously they generated errors)

# pct 0.2.5

- Fixed issue due to government data provider endpoint being down: https://github.com/ITSLeeds/pct/issues/51

# pct 0.2.4

- Updated vignettes use `tmap` instead of leaflet for easy-to-type map making code

# pct 0.2.3

- Remove OD pairs with no matching IDs, see https://github.com/ITSLeeds/pct/issues/47

# pct 0.2.2

- Updated training materials
- Fixed bug in `get_od()`

# pct 0.2.1

- `get_od()` now gets national data by default

# pct 0.2.0

- Updates to allow od data to be downloaded for `pct_regions`: see https://github.com/ITSLeeds/pct/issues/44

# pct 0.1.3

- Fix issue with `rnet` downloads - see https://github.com/ITSLeeds/pct/issues/45

# pct 0.1.2

- Bug fix: `get_pct_centroids()` and `get_pct_zones()` now work as intended
- Minor updates to vignettes

# pct 0.1.1
- Minor CRAN doi

# pct 0.1.0
* First release.


R/uptake.R
#' Calculate cycling uptake for UK 'Government Target' scenario
#'
#' Uptake model that takes distance and hilliness and returns
#' a percentage of trips that could be made by cycling along a desire line
#' under scenarios of change. Source: appendix of pct paper, hosted at:
#' [www.jtlu.org](https://www.jtlu.org/index.php/jtlu/article/download/862/1381/4359)
#' which states that: "To  estimate  cycling  potential,the  Propensity  to  Cycle  Tool  (PCT)  was
#' designed  to  use  the  best available  geographically  disaggregated
#' data  sources  on  travel  patterns."
#'
#' The functional form of the cycling uptake model used in the PCT is as follows:
#' (Source: [npct.github.io](https://npct.github.io/pct-shiny/regions_www/www/static/03a_manual/pct-bike-eng-user-manual-c1.pdf))
#'
#' ```
#' logit (pcycle) = -3.959 +   # alpha
#'   (-0.5963 * distance) +    # d1
#'   (1.866 * distancesqrt) +  # d2
#'   (0.008050 * distancesq) + # d3
#'   (-0.2710 * gradient) +    # h1
#'   (0.009394 * distance * gradient) +  # i1
#'   (-0.05135 * distancesqrt *gradient) # i2
#'
#' pcycle = exp ([logit (pcycle)]) / (1 + (exp([logit(pcycle)])
#' ```
#'
#' `uptake_pct_govtarget_2020()` and
#' `uptake_pct_godutch_2020()`
#' approximate the uptake models used in the updated 2020 release of
#' the PCT results.
#'
#' If the `distance` parameter is greater than 100, it is assumed that it is in m.
#' If for some reason you want to model cycling uptake associated with trips with
#' distances of less than 100 m, convert the distances to km first.
#'
#' @param distance Vector distance numeric values of routes in km
#' (switches to km if more than 100).
#' @param gradient Vector gradient numeric values of routes.
#' @param alpha The intercept
#' @param d1 Distance term 1
#' @param d2 Distance term 2
#' @param d3 Distance term 3
#' @param h1 Hilliness term 1
#' @param h2 Hilliness term 2
#' @param i1 Distance-hilliness interaction term 1
#' @param i2 Distance-hilliness interaction term 2
#' @param verbose Print messages? `FALSE` by default.
#'
#' @export
#' @examples
#' distance = 15
#' gradient = 2
#' logit_pcycle = -3.959 + # alpha
#'   (-0.5963 * distance) + # d1
#'   (1.866 * sqrt(distance)) + # d2
#'   (0.008050 * distance^2) + # d3
#'   (-0.2710 * gradient) + # h1
#'   (0.009394 * distance * gradient) + # i1
#'   (-0.05135 * sqrt(distance) * gradient) # i2
#' boot::inv.logit(logit_pcycle)
#' uptake_pct_govtarget(15, 2)
#' l = routes_fast_leeds
#' pcycle_scenario = uptake_pct_govtarget(l$length, l$av_incline)
#' pcycle_scenario_2020 = uptake_pct_govtarget_2020(l$length, l$av_incline)
#' plot(l$length, pcycle_scenario, ylim = c(0, 0.2))
#' points(l$length, pcycle_scenario_2020, col = "blue")
#'
#' # compare with published PCT data:
#' \dontrun{
#' l_pct_2020 = get_pct_lines(region = "isle-of-wight")
#' # test for another region:
#' # l_pct_2020 = get_pct_lines(region = "west-yorkshire")
#' l_pct_2020$rf_avslope_perc[1:5]
#' l_pct_2020$rf_dist_km[1:5]
#' govtarget_slc = uptake_pct_govtarget(
#'   distance = l_pct_2020$rf_dist_km,
#'   gradient = l_pct_2020$rf_avslope_perc
#' ) * l_pct_2020$all + l_pct_2020$bicycle
#' govtarget_slc_2020 = uptake_pct_govtarget_2020(
#'   distance = l_pct_2020$rf_dist_km,
#'   gradient = l_pct_2020$rf_avslope_perc
#' ) * l_pct_2020$all + l_pct_2020$bicycle
#' mean(l_pct_2020$govtarget_slc)
#' mean(govtarget_slc)
#' mean(govtarget_slc_2020)
#' godutch_slc = uptake_pct_godutch(
#'   distance = l_pct_2020$rf_dist_km,
#'   gradient = l_pct_2020$rf_avslope_perc
#' ) * l_pct_2020$all + l_pct_2020$bicycle
#' godutch_slc_2020 = uptake_pct_godutch_2020(
#'   distance = l_pct_2020$rf_dist_km,
#'   gradient = l_pct_2020$rf_avslope_perc
#' ) * l_pct_2020$all + l_pct_2020$bicycle
#' mean(l_pct_2020$dutch_slc)
#' mean(godutch_slc)
#' mean(godutch_slc_2020)
#' }
uptake_pct_govtarget = function(distance,
                                gradient,
                                alpha = -3.959,
                                d1 = -0.5963,
                                d2 = 1.866,
                                d3 = 0.008050,
                                h1 = -0.2710,
                                i1 = 0.009394,
                                i2 = -0.05135,
                                verbose = FALSE) {
  distance_gradient = check_distance_gradient(distance, gradient, verbose)
  distance = distance_gradient$distance
  gradient = distance_gradient$gradient
  pcycle_scenario = alpha +
    (d1 * distance) + # d1
    (d2 * sqrt(distance)) + # d2
    (d3 * distance ^ 2) + # d3
    (h1 * gradient) + # h1
    (i1 * distance * gradient) + # i1
    (i2 * sqrt(distance) * gradient) # i2
  boot::inv.logit(pcycle_scenario)
}

#' Calculate cycling uptake for UK 'Go Dutch' scenario
#'
#' This function implements the uptake model described in the original
#' Propensity to Cycle Tool paper (Lovelace et al. 2017):
#' https://doi.org/10.5198/jtlu.2016.862
#'
#' See [uptake_pct_govtarget()].
#'
#' @inheritParams uptake_pct_govtarget
#' @export
#' @examples
#' # https://www.jtlu.org/index.php/jtlu/article/download/862/1381/4359
#' # Equation 1B:
#' distance = 15
#' gradient = 2
#' logit = -3.959 + 2.523 +
#'   ((-0.5963 - 0.07626) * distance) +
#'   (1.866 * sqrt(distance)) +
#'   (0.008050 * distance^2) +
#'   (-0.2710 * gradient) +
#'   (0.009394 * distance * gradient) +
#'   (-0.05135 * sqrt(distance) * gradient)
#' logit
#' # Result: -3.144098
#'
#' pcycle = exp(logit) / (1 + exp(logit))
#' # Result: 0.04132445
#' boot::inv.logit(logit)
#' uptake_pct_godutch(distance, gradient,
#'   alpha = -3.959 + 2.523, d1 = -0.5963 - 0.07626,
#'   d2 = 1.866, d3 = 0.008050, h1 = -0.2710, i1 = 0.009394, i2 = -0.05135
#' )
#' # these are the default values
#' uptake_pct_godutch(distance, gradient)
#' l = routes_fast_leeds
#' pcycle_scenario = uptake_pct_godutch(l$length, l$av_incline)
#' plot(l$length, pcycle_scenario)
uptake_pct_godutch = function(distance,
                              gradient,
                              alpha = -3.959 + 2.523,
                              d1 = -0.5963 - 0.07626,
                              d2 = 1.866,
                              d3 = 0.008050,
                              h1 = -0.2710,
                              i1 = 0.009394,
                              i2 = -0.05135,
                              verbose = FALSE) {
  distance_gradient = check_distance_gradient(distance, gradient, verbose)
  distance = distance_gradient$distance
  gradient = distance_gradient$gradient
  logit_pcycle = alpha + (d1 * distance) +
    (d2 * sqrt(distance)) + (d3 * distance ^ 2) +
    (h1 * gradient) +
    (i1 * distance * gradient) +
    (i2 * sqrt(distance) * gradient)
  boot::inv.logit(logit_pcycle)
}

#' @rdname uptake_pct_govtarget
#' @export
uptake_pct_govtarget_2020 = function(distance,
                                     gradient,
                                     alpha = -4.018,
                                     d1 = -0.6369,
                                     d2 = 1.988,
                                     d3 = 0.008775,
                                     h1 = -0.2555,
                                     h2 = -0.78,
                                     i1 = 0.02006,
                                     i2 = -0.1234,
                                     verbose = FALSE) {
  distance_gradient = check_distance_gradient(distance, gradient, verbose)
  distance = distance_gradient$distance
  gradient = distance_gradient$gradient
  # Uptake formula from
  # https://raw.githubusercontent.com/npct/pct-shiny/
  # a59ebd1619af4400eeb7ffb2a8ecdd8ce4c3753d/
  # regions_www/www/static/03a_manual/pct-bike-eng-user-manual-c1.pdf
  #
  # logit (pcycle)= -4.018 +  (-0.6369 *  distance)  +
  #   (1.988  * distancesqrt)  +  (0.008775* distancesq) +
  #   (-0.2555* gradient) + (0.02006* distance*gradient) +
  #   (-0.1234* distancesqrt*gradient)
  gradient = gradient + h2
  pcycle_scenario = alpha +
    (d1 * distance) + # d1
    (d2 * sqrt(distance)) + # d2
    (d3 * distance ^ 2) + # d3
    (h1 * gradient) + # h1
    (i1 * distance * gradient) + # i1
    (i2 * sqrt(distance) * gradient) # i2
  boot::inv.logit(pcycle_scenario)
}

#' @rdname uptake_pct_govtarget
#' @export
uptake_pct_godutch_2020 = function(distance,
                                   gradient,
                                   alpha = -4.018 + 2.550,
                                   d1 = -0.6369 - 0.08036,
                                   d2 = 1.988,
                                   d3 = 0.008775,
                                   h1 = -0.2555,
                                   h2 = -0.78,
                                   i1 = 0.02006,
                                   i2 = -0.1234,
                                   verbose = FALSE) {
  distance_gradient = check_distance_gradient(distance, gradient, verbose)
  distance = distance_gradient$distance
  gradient = distance_gradient$gradient
  # Uptake formula from manual:
  # logit_pcycle = -4.018  +  (-0.6369  *  distance)  +  (1.988  *  distancesqrt) +
  # (0.008775  * distancesq) + (-0.2555 * gradient) + (0.02006 * distance*gradient) +
  # (-0.1234 * distancesqrt*gradient) + (2.550 * dutch) +  (-0.08036* dutch * distance) +
  # (0.05509* ebike * distance) + (-0.0002950* ebike * distancesq) + (0.1812* ebike * gradient)
  gradient = gradient + h2
  pcycle_scenario = alpha +
    (d1 * distance) + # d1
    (d2 * sqrt(distance)) + # d2
    (d3 * distance ^ 2) + # d3
    (h1 * gradient) + # h1
    (i1 * distance * gradient) + # i1
    (i2 * sqrt(distance) * gradient) # i2
  boot::inv.logit(pcycle_scenario)
}

#' @rdname uptake_pct_govtarget
#' @export
uptake_pct_ebike_2020 = function(distance,
                                   gradient,
                                   alpha = -4.018 + 2.550,
                                   d1 = -0.6369 - 0.08036 + 0.05509,
                                   d2 = 1.988,
                                   d3 = 0.008775 -0.0002950,
                                   h1 = -0.2555 + 0.1812,
                                   h2 = -0.78,
                                   i1 = 0.02006,
                                   i2 = -0.1234,
                                   verbose = FALSE) {
  distance_gradient = check_distance_gradient(distance, gradient, verbose)
  distance = distance_gradient$distance
  gradient = distance_gradient$gradient
  # Uptake formula from manual:
  # logit_pcycle = -4.018  +  (-0.6369  *  distance)  +  (1.988  *  distancesqrt) +
  # (0.008775  * distancesq) + (-0.2555 * gradient) + (0.02006 * distance*gradient) +
  # (-0.1234 * distancesqrt*gradient) + (2.550 * dutch) +  (-0.08036* dutch * distance) +
  # (0.05509* ebike * distance) + (-0.0002950* ebike * distancesq) + (0.1812* ebike * gradient)
  gradient = gradient + h2
  pcycle_scenario = alpha +
    (d1 * distance) + # d1
    (d2 * sqrt(distance)) + # d2
    (d3 * distance ^ 2) + # d3
    (h1 * gradient) + # h1
    (i1 * distance * gradient) + # i1
    (i2 * sqrt(distance) * gradient) # i2
  boot::inv.logit(pcycle_scenario)
}

#' @rdname uptake_pct_govtarget
#' @export
#' @examples
#' # Take an origin destination (OD) pair between an LSOA centroid and a
#' # secondary school. In this OD pair, 30 secondary school children travel, of
#' # whom 3 currently cycle. The fastest route distance is 3.51 km and the
#' # gradient is 1.11%. The
#' # gradient as centred on Dutch hilliness levels is 1.11 – 0.63 = 0.48%.
#' # The observed number of cyclists is 2. ... Modelled baseline= 30 * .0558 = 1.8.
#' uptake_pct_govtarget_school2(3.51, 1.11)
uptake_pct_govtarget_school2 = function(
  distance,
  gradient,
  alpha = -7.178,
  d1 = -1.870,
  d2 = 5.961,
  # d3 = -0.2401,
  h1 = -0.5290,
  h2 = -0.63,
  verbose = FALSE
  # i1 = 0.02006,
  # i2 = -0.1234
) {
  # Uptake formula from Goodman et al. (2019)
  # http://www.sciencedirect.com/science/article/pii/S2214140518301257
  #
  # Equation 1.1 (primary school children):
  #   logit (pcycle) = −4.813 + (0.9743 * distance) + (−0.2401 * distancesq) + (−0.4245 * centred_gradient)
  # pcycle = exp ([logit (pcycle)])/(1 + (exp([logit(pcycle)]))).
  # Equation 2.1 (secondary school children):
  #   logit (pcycle) = −7.178 + (−1.870 * distance) + (5.961 * distance sqrt) + (−0.5290 * centred_gradient)
  # pcycle = exp ([logit (pcycle)])/(1 + (exp([logit(pcycle)])))
  distance_gradient = check_distance_gradient(distance, gradient, verbose)
  distance = distance_gradient$distance
  gradient = distance_gradient$gradient
  gradient = gradient + h2
  pcycle_scenario = alpha +
    (d1 * distance) + # d1
    (d2 * sqrt(distance)) + # d2
    # (d3 * distance^2) + # d3
    (h1 * gradient)
  # +    # h1
  # (i1 * distance * gradient) +  # i1
  # (i2 * sqrt(distance) * gradient) # i2
  boot::inv.logit(pcycle_scenario)
}

#' @rdname uptake_pct_govtarget
#' @export
#' @examples
#' # pcycle = exp ([logit (pcycle)])/(1 + (exp([logit(pcycle)]))).
#' # pcycle = exp(1.953)/(1 + exp(1.953)) = .8758, or 87.58%.
#' uptake_pct_godutch_school2(3.51, 1.11)
uptake_pct_godutch_school2 = function(
  distance,
  gradient,
  alpha = -7.178 + 3.574,
  d1 = -1.870 + 0.3438,
  d2 = 5.961,
  # d3 = -0.2401,
  h1 = -0.5290,
  h2 = -0.63,
  verbose = FALSE
  # i1 = 0.02006,
  # i2 = -0.1234
) {
  # Uptake formula from Goodman et al. (2019)
  # http://www.sciencedirect.com/science/article/pii/S2214140518301257
  #
  # Equation 2.2 (secondary school children):
  #   logit(pcycle) = Equation 1.2 + Dutch parameters.
  # logit (pcycle) = −7.178 + (−1.870 * distance) + (5.961 * distancesqrt) + (−0.5290 * centred_gradient)+(3.574) +
  #   (0.3438 * distance).
  # pcycle = exp ([logit (pcycle)])/(1 + (exp([logit(pcycle)]))).
  #

  # logit (pcycle)= -4.018 +  (-0.6369 *  distance)  +
  #   (1.988  * distancesqrt)  +  (0.008775* distancesq) +
  #   (-0.2555* gradient) + (0.02006* distance*gradient) +
  #   (-0.1234* distancesqrt*gradient)
  distance_gradient = check_distance_gradient(distance, gradient, verbose)
  distance = distance_gradient$distance
  gradient = distance_gradient$gradient
  gradient = gradient + h2
  pcycle_scenario = alpha +
    (d1 * distance) + # d1
    (d2 * sqrt(distance)) + # d2
    # (d3 * distance^2) + # d3
    (h1 * gradient)
  # +    # h1
  # (i1 * distance * gradient) +  # i1
  # (i2 * sqrt(distance) * gradient) # i2
  boot::inv.logit(pcycle_scenario)
}

check_distance_gradient = function(distance, gradient, verbose = TRUE) {
  if (!is.numeric(c(distance, gradient))) {
    stop("distance and gradient need to be numbers.")
  }
  # is it in m
  if (mean(distance, na.rm = TRUE) > 100) {
    if (verbose) {
      message("Distance assumed in m, switching to km")
    }
    distance = distance / 1000
  }
  # If distance is more than 30 km set to 30:
  which_distance_too_long = distance > 30
  if (any(which_distance_too_long)) {
    if (verbose) {
      message("Distances greater than 30 km set to 30 km")
    }
    distance[which_distance_too_long] = 30
  }
  # is it in %? If mean of gradient is more than 0.1 (10% as ratio), probably
  is_gradient_percent = mean(gradient) > 0.1
  if(!is_gradient_percent) {
    if (verbose) {
      message("Gradient assumed to be gradient, switching to % (*100)")
    }
    gradient = gradient * 100
  }
  list(distance = distance, gradient = gradient)
}
